Add section to tidyverse module to explicitly cover matrices (#234)
AlexsLemonade/training-modules: intro-to-R-tidyverse/03-intro_to_tidyverse.Rmd

pick a type that it can convert everything to.
What does it choose?

```{r matrix}
# Coerce `gene_df` into a matrix
gene_matrix <- as.matrix(gene_df)
```

```{r matrix-type, live = TRUE}
# Explore the structure of the `gene_matrix` object
str(gene_matrix)
```

While that worked, it is rare that we want numbers converted to text, so we are
going to select only the columns with numeric values before converting it to a matrix.
We can do this most easily by removing the first column, which contains the gene names 
stored as character values.

```{r matrix-numeric, live = TRUE}
# Let's save a new matrix object names `gene_num_matrix` containing only
# the numeric values
gene_num_matrix <- as.matrix(gene_df[, -1])

# Explore the structure of the `gene_num_matrix` object
str(gene_num_matrix)
```

Why do we have a `[, -1]` after `gene_df` in the above chunk?

Now that the matrix is all numbers, we can do things like calculate the column
or row statistics using `apply()`.

```{r rowmeans}
# Calculate row means
gene_means <- apply(gene_num_matrix, 1, mean) # Notice we are using 1 here

# How long will `gene_means` be? 
length(gene_means)
```

> Note that we can obtain the same results using the columns with numeric values
> from the `gene_df` data frame.
> 
> ```{r rowmeans-dataframe}
> # Calculate row means using the `gene_df` object
> gene_means_from_df <- apply(gene_df[, -1], 1, mean) # Removing the character
> # column allows R to do the as.matrix() coercion automatically, which only works
> # if the rest of the data frame is numeric.

Reviewer: ```suggestion
Note that we can obtain the same results if we select just the columns with numeric
values from the `gene_df` data frame.
This allows R to do the as.matrix() coercion automatically, and can be a handy shortcut
if you have a *mostly* numeric data frame. 

```{r rowmeans-dataframe}
# Calculate row means using the `gene_df` object after removing the character column
# apply() converts this to a matrix internally
gene_means_from_df <- apply(gene_df[, -1], 1, mean) 
```

performing transformations across a data frame, matrix or other object structures. 

One of this family is (shockingly) the function `apply()`, which operates on 
matrices.

A matrix is similar to a data frame in that it is a rectangular table of data,
but it has an additional constraint: 
rather than each column having a type, ALL data in a matrix has the same type.

The first argument to `apply()` is the data object we want to work on.
The third argument is the function we will apply to each row or column of the 
data object.
The second argument in specifies whether we are applying the function 
across rows or across columns (1 for rows, 2 for columns).

Remember that `gene_df` is a gene x sample gene expression data frame that has
columns of two different types, character and numeric, but converting it to a
matrix will require us to make them all the same type.

> We can still coerce it into a matrix using `as.matrix()`, in which case R will

Reviewer: ```suggestion
We can coerce it into a matrix using `as.matrix()`, in which case R will
```

pick a type that it can convert everything to.
What does it choose?

```{r matrix}
# Coerce `gene_df` into a matrix
gene_matrix <- as.matrix(gene_df)
```

```{r matrix-type, live = TRUE}
# Explore the structure of the `gene_matrix` object
str(gene_matrix)
```

While that worked, it is rare that we want numbers converted to text, so we are

> going to select only the columns with numeric values.

Reviewer: ```suggestion
going to select only the columns with numeric values before converting it to a matrix.
```

> We can do this most easily by removing the first column, which contains character
> values.

Reviewer: ```suggestion
We can do this most easily by removing the first column, which contains the gene names 
stored as character values.
```

functions (`tapply()`, `lapply()`, etc.) but you can read more information about 
them [here](https://www.guru99.com/r-apply-sapply-tapply.html).

## Matrices

> A matrix might at first appear similar to a data frame, but it has an additional
> constraint: rather than each column having a single type, ALL data in the matrix
> has the same type.

Reviewer: ```suggestion
A matrix is similar to a data frame in that it is a rectangular table of data, but it has an additional constraint: 
rather than each column having a type, ALL data in a matrix has the same type.
```

> The data frame above is made up of two different types, character and numeric,
> so converting it to a matrix would require us to make it all the same type.

Reviewer: ```suggestion
The data frame above has  columns of two different types, character and numeric, but converting it to a matrix will require us to make them all the same type.
```

> If we try to coerce it into a matrix using `as.matrix()`, R will have to pick a
> type that it can convert everything to.

Reviewer: ```suggestion
We can still coerce it into a matrix using `as.matrix()`, in which case R will pick a type that it can convert everything to.
```

We can do this most easily by removing the first column, as we did in `apply()`
function examples above.

```{r matrix-numeric, live = TRUE}
# Let's save a new matrix object names `gene_num_matrix` containing only
# the numeric values
gene_num_matrix <- as.matrix(gene_df[, -1])

# Explore the structure of the `gene_num_matrix` object
str(gene_num_matrix)
```

Now that the matrix is all numbers, we can do things like calculate the column
or row statistics using `apply()` as we did above.

```{r matrix-rowmeans}
# Calculate row means for our numeric matrix object
gene_matrix_means <- apply(gene_num_matrix, 1, mean)
length(gene_matrix_means)
str(gene_matrix_means)
```

Note that the length and the structure of the `gene_matrix_means` object is the
same as the length and structure of the `gene_means` object which we generated
using the `gene_df` data frame.

The difference here is that the `gene_means` data frame object would contain a
combination of character and numeric data types if we did not remove the first
column, while the `gene_matrix_means` object would contain only one data type,
character, if we did not remove the first column because we coerced the object

> into a matrix.

Reviewer: This isn't quite accurate, as R will convert to a the data frame to a matrix before using `apply()`. Apply only really works on matrices. So the result will be trying to calculate `mean()` on a character vectors, which will give `NA`s.

> We can put the character gene names back into the numeric matrix object by

Reviewer: ```suggestion
We can put the gene names back into the numeric matrix object by
```

assigning them as rownames.

```{r matrix-rownames, live = TRUE}
# Assign the gene names from gene_df$Gene to the `gene_num_matrix` object using
# the `rownames()` function
rownames(gene_num_matrix) <- gene_df$Gene

# Explore the `gene_num_matrix` object
head(gene_num_matrix)
```

> 

Reviewer: I liked my comment about row names from the muddiest point nb, so I am going to suggest adding it in here as well: 
"Row names like this can be very convenient for keeping matrices organized, but row names (and column names) can be lost or misordered if you are not careful, especially during input and output, so treat them with care."
